gordon_janitor_gcp/plugins/authority.py

```diff
@@ -0,0 +1,192 @@
+
+    from gordon_janitor_gcp import plugins
+
+    async def run():
+        rrset_channel = asyncio.queue()
+        authority = plugins.get_authority(config, rrset_channel)
+        await authority.start()
+        msg = await rrset_channel.get()
+        print(msg)
+
+    loop = asyncio.get_event_loop()
+    loop.run_until_complete(run())
+    # prints: {'zone': 'example.com', 'resourceRecords': [...]}
+"""
+
+import asyncio
+import logging
+
+import aiohttp
+import zope.interface
+from gordon_janitor import interfaces
+
+from gordon_janitor_gcp import exceptions
+from gordon_janitor_gcp.clients import auth
+from gordon_janitor_gcp.clients import gce
+from gordon_janitor_gcp.clients import gcrm
+
+
+class GCEAuthorityBuilder:
+    """Build and configure a :class:`GCEAuthority` object.
+
+    Args:
+        config (dict): plugin-specific configuration.
+        rrset_channel (asyncio.Queue): channel to send resource record messages
+            to.
+    """
+    def __init__(self, config, rrset_channel, **kwargs):
+        self.config = config
+        self.rrset_channel = rrset_channel
+        self.kwargs = kwargs
+        self._loop = asyncio.get_event_loop()
+        self.session = aiohttp.ClientSession(loop=self._loop)
+
+    def _get_crm_client(self, keyfile_path, scopes):
+        crm_auth = auth.GAuthClient(
+            keyfile_path, scopes=scopes, session=self.session)
+        return gcrm.GCRMClient(crm_auth, self.session)
+
+    def _get_gce_client(self, keyfile_path, scopes):
+        tag_blacklist = self.config.get('tag_blacklist', [])
+        _metadata_blacklist = self.config.get('metadata_blacklist', [])
+        metadata_blacklist = [dict([pair]) for pair in _metadata_blacklist]
+
+        gce_auth = auth.GAuthClient(keyfile_path, scopes=scopes,
+                                    session=self.session)
+        return gce.GCEClient(gce_auth, self.session,
+                             blacklisted_tags=tag_blacklist,
+                             blacklisted_metadata=metadata_blacklist)
+
+    def _validate_config(self):
+        if not self.config.get('keyfile'):
+            msg = ('The path to a Service Account JSON keyfile is required to '
+                   'authenticate to Google Compute Engine and Cloud '
+                   'Resource Manager.')
+            logging.error(msg)
+            raise exceptions.GCPConfigError(msg)
+        if not self.config.get('zone'):
+            msg = ('The ID of the target Cloud DNS managed zone is required to '
+                   'identify which zone records should belong to.')
+            logging.error(msg)
+            raise exceptions.GCPConfigError(msg)
+
+    def build_authority(self):
+        """Set up dependencies and create a GCEAuthority object.
+
+        Returns:
+            obj: A configured GCEAuthority object.
+        """
+        try:
+            self._validate_config()
+        except exceptions.GCPConfigError:
+            self.session.close()
```

Comment:
It may have been lost in your most recent push, but I think setting the `self.session` object after you validate config would be better. (I also don't think you're properly re-raising the config error) i.e:

```py
class GCEAuthorityBuilder:
    def __init__(...):
        self.session = None

    def build_authority(self):
        self._validate_config()  # will raise on its own, no need to catch + reraise
        self.session = aiohttp.ClientSession()
        # continue on
```